
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Code accompanying the iicd workshop 2024">
      
      
        <meta name="author" content="Achille Nazaret">
      
      
        <link rel="canonical" href="https://anazaret.github.io/iicd-workshop-2024/problem_set/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../references/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.28">
    
    
      
        <title>Problem set - iicd-workshop-2024</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#problem-set-3-probabilistic-modeling" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="iicd-workshop-2024" class="md-header__button md-logo" aria-label="iicd-workshop-2024" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            iicd-workshop-2024
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Problem set
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="deep-orange"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="deep-orange"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/anazaret/iicd-workshop-2024" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    anazaret/iicd-workshop-2024
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="iicd-workshop-2024" class="md-nav__button md-logo" aria-label="iicd-workshop-2024" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    iicd-workshop-2024
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/anazaret/iicd-workshop-2024" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    anazaret/iicd-workshop-2024
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Problem set
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Problem set
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#problem-1-simple-model-with-global-gene-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 1 - Simple model with global gene parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Problem 1 - Simple model with global gene parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-load-the-data" class="md-nav__link">
    <span class="md-ellipsis">
      1) Load the data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-implement-a-simple-normal-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      2) Implement a simple Normal gene model
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2) Implement a simple Normal gene model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-implement-the-__init__-and-the-get_distribution-methods" class="md-nav__link">
    <span class="md-ellipsis">
      2.1) Implement the __init__ and the get_distribution methods
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-fit-the-model" class="md-nav__link">
    <span class="md-ellipsis">
      2.2) Fit the model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-visualize-the-learned-means" class="md-nav__link">
    <span class="md-ellipsis">
      2.3) Visualize the learned means
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-visualize-a-few-gene-distributions" class="md-nav__link">
    <span class="md-ellipsis">
      2.4) Visualize a few gene distributions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-now-repeat-the-same-steps-for-a-poisson-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      3) Now repeat the same steps for a Poisson gene model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-now-repeat-the-same-steps-for-a-negative-binomial-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      4) Now repeat the same steps for a Negative Binomial gene model
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#problem-2-gene-model-with-cell-specific-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 2 - Gene model with cell specific parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Problem 2 - Gene model with cell specific parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-implement-the-auto-encoder-model" class="md-nav__link">
    <span class="md-ellipsis">
      1) Implement the auto-encoder model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-fit-the-model" class="md-nav__link">
    <span class="md-ellipsis">
      2) Fit the model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-implement-a-get_latent_representation-method" class="md-nav__link">
    <span class="md-ellipsis">
      3) Implement a get_latent_representation method
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-visualize-the-learned-cell-specific-representations-using-umap" class="md-nav__link">
    <span class="md-ellipsis">
      4) Visualize the learned cell specific representations using UMAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-compare-against-decipher" class="md-nav__link">
    <span class="md-ellipsis">
      5) Compare against Decipher
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#problem-3-implement-your-own-model" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 3 - Implement your own model
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../references/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#problem-1-simple-model-with-global-gene-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 1 - Simple model with global gene parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Problem 1 - Simple model with global gene parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-load-the-data" class="md-nav__link">
    <span class="md-ellipsis">
      1) Load the data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-implement-a-simple-normal-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      2) Implement a simple Normal gene model
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2) Implement a simple Normal gene model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-implement-the-__init__-and-the-get_distribution-methods" class="md-nav__link">
    <span class="md-ellipsis">
      2.1) Implement the __init__ and the get_distribution methods
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-fit-the-model" class="md-nav__link">
    <span class="md-ellipsis">
      2.2) Fit the model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-visualize-the-learned-means" class="md-nav__link">
    <span class="md-ellipsis">
      2.3) Visualize the learned means
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-visualize-a-few-gene-distributions" class="md-nav__link">
    <span class="md-ellipsis">
      2.4) Visualize a few gene distributions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-now-repeat-the-same-steps-for-a-poisson-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      3) Now repeat the same steps for a Poisson gene model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-now-repeat-the-same-steps-for-a-negative-binomial-gene-model" class="md-nav__link">
    <span class="md-ellipsis">
      4) Now repeat the same steps for a Negative Binomial gene model
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#problem-2-gene-model-with-cell-specific-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 2 - Gene model with cell specific parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Problem 2 - Gene model with cell specific parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-implement-the-auto-encoder-model" class="md-nav__link">
    <span class="md-ellipsis">
      1) Implement the auto-encoder model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-fit-the-model" class="md-nav__link">
    <span class="md-ellipsis">
      2) Fit the model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-implement-a-get_latent_representation-method" class="md-nav__link">
    <span class="md-ellipsis">
      3) Implement a get_latent_representation method
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-visualize-the-learned-cell-specific-representations-using-umap" class="md-nav__link">
    <span class="md-ellipsis">
      4) Visualize the learned cell specific representations using UMAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-compare-against-decipher" class="md-nav__link">
    <span class="md-ellipsis">
      5) Compare against Decipher
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#problem-3-implement-your-own-model" class="md-nav__link">
    <span class="md-ellipsis">
      Problem 3 - Implement your own model
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="problem-set-3-probabilistic-modeling">Problem set 3 - Probabilistic modeling<a class="headerlink" href="#problem-set-3-probabilistic-modeling" title="Permanent link">&para;</a></h1>
<p>Wednesday, 2024-07-10</p>
<hr />
<p>In this problem set, we explore probabilistic modeling on scRNA-sequencing data.</p>
<p>You can find the solutions to this problem set in this notebook: <a href="https://github.com/ANazaret/iicd-workshop-2024/blob/main/problem-set-3.ipynb">problem_set_3.ipynb</a>.</p>
<p>We recommend starting a new conda environment for this problem set.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>iicd-workshop<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.11
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>conda<span class="w"> </span>activate<span class="w"> </span>iicd-workshop
</span></code></pre></div></p>
<p>Before starting, make sure the <code>iicd-workshop-2024</code> package is installed, and that you have the
latest version of it.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>pip<span class="w"> </span>install<span class="w"> </span>iicd-workshop-2024<span class="w"> </span>--upgrade
</span></code></pre></div></p>
<p>As well as other dependencies:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>pip<span class="w"> </span>install<span class="w"> </span>scvi-tools<span class="w"> </span>torch<span class="w"> </span>scanpy<span class="w"> </span>seaborn<span class="w"> </span>matplotlib
</span></code></pre></div></p>
<h2 id="problem-1-simple-model-with-global-gene-parameters">Problem 1 - Simple model with global gene parameters<a class="headerlink" href="#problem-1-simple-model-with-global-gene-parameters" title="Permanent link">&para;</a></h2>
<p>In this problem, you will implement simple models with global gene parameters.
For each cell <span class="arithmatex">\(i\)</span> and gene <span class="arithmatex">\(g\)</span>, assume that the gene expression <span class="arithmatex">\(x_{i,g}\)</span> follows a distribution
$$ x_{i,g} \sim p(\theta_g) $$
where <span class="arithmatex">\(\theta_g\)</span> contains gene specific parameters shared across cells, and <span class="arithmatex">\(p\)</span> is a distribution (e.g. Normal, Negative Binomial).</p>
<h3 id="1-load-the-data">1) Load the data<a class="headerlink" href="#1-load-the-data" title="Permanent link">&para;</a></h3>
<p>Load public scRNA-seq data <code>pbmc</code> from the <code>scvi-tools</code> package.
<div class="language-python highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="kn">import</span> <span class="nn">scvi</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="n">adata</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">pbmc_dataset</span><span class="p">()</span>
</span></code></pre></div></p>
<h3 id="2-implement-a-simple-normal-gene-model">2) Implement a simple Normal gene model<a class="headerlink" href="#2-implement-a-simple-normal-gene-model" title="Permanent link">&para;</a></h3>
<p>In this question, assume that
$$ x_{i,g} \sim \mathcal{N}(\mu_g, \sigma_g^2) $$
where <span class="arithmatex">\(\mu_g\)</span> and <span class="arithmatex">\(\sigma_g\)</span> are gene specific parameters shared across cells.</p>
<p>You will implement simple gene model by subclassing the <code>BaseGeneModel</code> class provided
in the <code>iicd_workshop_2024</code> package. Link to the documentation: <a href="../references/#iicd_workshop_2024.gene_model.BaseGeneModel">BaseGeneModel</a>
This class already implements the <code>fit</code> and <code>loss</code> methods for you.
You just need to implement the <code>get_distribution</code> method that returns the distribution of the model,
as well as the <code>__init__</code> method that initializes the model.</p>
<p>Here is a template for the <code>NormalGeneModel</code> class:
<div class="language-python highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="kn">import</span> <span class="nn">torch.distributions</span> <span class="k">as</span> <span class="nn">dist</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="kn">from</span> <span class="nn">iicd_workshop_2024.gene_model</span> <span class="kn">import</span> <span class="n">BaseGeneModel</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>
</span><span id="__span-4-5"><a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="k">class</span> <span class="nc">NormalGeneModel</span><span class="p">(</span><span class="n">BaseGeneModel</span><span class="p">):</span>
</span><span id="__span-4-6"><a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="__span-4-7"><a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">n_genes</span><span class="p">)</span>
</span><span id="__span-4-8"><a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>        <span class="c1"># declare any parameters here, using torch.nn.Parameter</span>
</span><span id="__span-4-9"><a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>        <span class="o">...</span>
</span><span id="__span-4-10"><a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a>
</span><span id="__span-4-11"><a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>    <span class="k">def</span> <span class="nf">get_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_idx</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dist</span><span class="o">.</span><span class="n">Distribution</span><span class="p">:</span>
</span><span id="__span-4-12"><a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>        <span class="c1"># return the distribution for the gene_idx-th gene</span>
</span><span id="__span-4-13"><a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>        <span class="k">return</span> <span class="o">...</span>
</span></code></pre></div></p>
<h4 id="21-implement-the-__init__-and-the-get_distribution-methods">2.1) Implement the <code>__init__</code> and the <code>get_distribution</code> methods<a class="headerlink" href="#21-implement-the-__init__-and-the-get_distribution-methods" title="Permanent link">&para;</a></h4>
<p>You should use the <code>dist.Normal</code> class from <code>torch.distributions</code> to create the distribution object.
Here is the documentation for <a href="https://pytorch.org/docs/stable/distributions.html">torch.distributions</a>.</p>
<h4 id="22-fit-the-model">2.2) Fit the model<a class="headerlink" href="#22-fit-the-model" title="Permanent link">&para;</a></h4>
<p>Fit the model to the data using the <code>fit</code> method of the <code>NormalGeneModel</code> class.
We recommend you to read the documentation and source code for the <a href="../references/#iicd_workshop_2024.inference.fit">fit</a> method.</p>
<h4 id="23-visualize-the-learned-means">2.3) Visualize the learned means<a class="headerlink" href="#23-visualize-the-learned-means" title="Permanent link">&para;</a></h4>
<p>Visualize the learned gene means against the true empirical gene means.</p>
<ul>
<li>Are they the same?</li>
<li>If not, can you hypothesize reasons for any differences?</li>
<li>And can you fix it?</li>
</ul>
<p>To visualize the gene means, you can use the following code snippet:
<div class="language-python highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a>
</span><span id="__span-5-5"><a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="k">def</span> <span class="nf">plot_learned_vs_empirical_mean</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">adata</span><span class="p">):</span>
</span><span id="__span-5-6"><a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a>    <span class="n">empirical_means</span> <span class="o">=</span> <span class="o">...</span>
</span><span id="__span-5-7"><a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a>    <span class="n">learned_means</span> <span class="o">=</span> <span class="o">...</span>
</span><span id="__span-5-8"><a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a>    <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">empirical_means</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">learned_means</span><span class="p">)</span>
</span><span id="__span-5-9"><a id="__codelineno-5-9" name="__codelineno-5-9" href="#__codelineno-5-9"></a>    <span class="n">max_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">empirical_means</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">learned_means</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
</span><span id="__span-5-10"><a id="__codelineno-5-10" name="__codelineno-5-10" href="#__codelineno-5-10"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">)</span>
</span><span id="__span-5-11"><a id="__codelineno-5-11" name="__codelineno-5-11" href="#__codelineno-5-11"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Empirical mean&quot;</span><span class="p">)</span>
</span><span id="__span-5-12"><a id="__codelineno-5-12" name="__codelineno-5-12" href="#__codelineno-5-12"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Learned mean&quot;</span><span class="p">)</span>
</span><span id="__span-5-13"><a id="__codelineno-5-13" name="__codelineno-5-13" href="#__codelineno-5-13"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</span></code></pre></div></p>
<h4 id="24-visualize-a-few-gene-distributions">2.4) Visualize a few gene distributions<a class="headerlink" href="#24-visualize-a-few-gene-distributions" title="Permanent link">&para;</a></h4>
<p>Visualize a few gene distributions by plotting the learned distributions against the empirical distribution.
For example, genes: <code>["CD14", "CD74", "RPS27"]</code>.</p>
<p>You can use the function <code>plot_gene_distribution</code> from the <code>iicd_workshop_2024.gene_model</code> module.
See documentation for <a href="../references/#iicd_workshop_2024.gene_model.plot_gene_distribution">plot_gene_distribution</a>.</p>
<h3 id="3-now-repeat-the-same-steps-for-a-poisson-gene-model">3) Now repeat the same steps for a Poisson gene model<a class="headerlink" href="#3-now-repeat-the-same-steps-for-a-poisson-gene-model" title="Permanent link">&para;</a></h3>
<p>Math reminder:</p>
<ul>
<li>The Poisson distribution is a distribution over non-negative integers.</li>
<li>It is parametrized by a single parameter <span class="arithmatex">\(\lambda\)</span>, its mean (which is also its variance).
$$ x \sim \text{Poisson}(\lambda) $$
with <span class="arithmatex">\(\mathbb{E}[x] = \text{Var}[x] = \lambda\)</span>.</li>
</ul>
<h3 id="4-now-repeat-the-same-steps-for-a-negative-binomial-gene-model">4) Now repeat the same steps for a Negative Binomial gene model<a class="headerlink" href="#4-now-repeat-the-same-steps-for-a-negative-binomial-gene-model" title="Permanent link">&para;</a></h3>
<p>Math reminder:</p>
<ul>
<li>The Negative Binomial distribution is a distribution over non-negative integers.</li>
<li>It is parametrized by two parameters: the mean <span class="arithmatex">\(\mu\)</span> and the inverse dispersion parameter <span class="arithmatex">\(\alpha\)</span>.
$$ x \sim \text{NegBinom}(\mu, \alpha) $$ with
$ \mathbb{E}[x] = \mu$ and <span class="arithmatex">\(\text{Var}[x] = \mu + \frac{\mu^2}{\alpha}\)</span>.</li>
<li>The Negative Binomial distribution implemented in pytorch does not use the mean and inverse dispersion parameter.
  Instead, it uses the <code>total_count</code> and <code>logit</code> parameters.
  The correspondance is given by:</li>
<li>total_count = inverse dispersion</li>
<li>logit = log(mean) - log(inverse dispersion)</li>
</ul>
<h2 id="problem-2-gene-model-with-cell-specific-parameters">Problem 2 - Gene model with cell specific parameters<a class="headerlink" href="#problem-2-gene-model-with-cell-specific-parameters" title="Permanent link">&para;</a></h2>
<p>In this problem, you will implement gene models with cell specific parameters, learned with an
autoencoder.</p>
<p>In the previous problem, we assumed that the gene expression <span class="arithmatex">\(x_{i,g}\)</span> was distributed according to:
$$ x_{i,g} \sim p(\theta_g) $$
where <span class="arithmatex">\(\theta_g\)</span> contains gene specific parameters shared across cells.
In reality, the gene expression is influenced by cell specific parameters as well.</p>
<p>The idea of representation learning for high-dimensional data is that we might not
know exactly what are the cell specific parameters that influence the gene expression,
but we assume that there exist a small number of them.</p>
<p>We write <span class="arithmatex">\(z_i\)</span> the cell specific representation and we obtain the model:
$$
\begin{align}
x_{i,g} &amp;\sim p(f(z_i), \theta_g)
\end{align}
$$
where:
<span class="arithmatex">\(f\)</span> is a function that transforms the cell specific representation into the gene specific parameters.</p>
<p>For this problem, we define the family of distributions <span class="arithmatex">\(p\)</span> to be negative binomial distributions</p>
<p>We further will use amortized inference to learn the cell specific representations <span class="arithmatex">\(z_i\)</span>.
$$ z_i \approx g(x_i), $$
where <span class="arithmatex">\(g\)</span> is a neural network that takes the gene expression <span class="arithmatex">\(x_i\)</span> as input and outputs the cell specific representation <span class="arithmatex">\(z_i\)</span>.</p>
<h3 id="1-implement-the-auto-encoder-model">1) Implement the auto-encoder model<a class="headerlink" href="#1-implement-the-auto-encoder-model" title="Permanent link">&para;</a></h3>
<p>Implement the auto-encoder model by completing the following class.
You may use the <a href="../references/#iicd_workshop_2024.inference.fit">fit</a> function to train the model
as is done in the <a href="../references/#iicd_workshop_2024.gene_model.BaseGeneModel">BaseGeneModel</a>.
<div class="language-python highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="kn">import</span> <span class="nn">torch</span>
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>
</span><span id="__span-6-4"><a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="k">class</span> <span class="nc">LatentModel</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
</span><span id="__span-6-5"><a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">n_latent</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="__span-6-6"><a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
</span><span id="__span-6-7"><a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>        <span class="o">...</span>
</span><span id="__span-6-8"><a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>
</span><span id="__span-6-9"><a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>    <span class="k">def</span> <span class="nf">get_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
</span><span id="__span-6-10"><a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>       <span class="o">...</span>
</span><span id="__span-6-11"><a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>
</span><span id="__span-6-12"><a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>    <span class="k">def</span> <span class="nf">loss</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
</span><span id="__span-6-13"><a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>        <span class="k">return</span> <span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">forward</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">log_prob</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</span></code></pre></div></p>
<p>You can also use the <a href="../references/#iicd_workshop_2024.neural_network.DenseNN">DenseNN</a>
class from the <code>iicd_workshop_2024.neural_network</code> module to define the neural network.
<div class="language-python highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="kn">from</span> <span class="nn">iicd_workshop_2024.neural_network</span> <span class="kn">import</span> <span class="n">DenseNN</span>
</span></code></pre></div></p>
<h3 id="2-fit-the-model">2) Fit the model<a class="headerlink" href="#2-fit-the-model" title="Permanent link">&para;</a></h3>
<p>You can use the <code>fit</code> function from the <code>iicd_workshop_2024.inference</code> module to fit the model.</p>
<h3 id="3-implement-a-get_latent_representation-method">3) Implement a <code>get_latent_representation</code> method<a class="headerlink" href="#3-implement-a-get_latent_representation-method" title="Permanent link">&para;</a></h3>
<p>This function should be able to retrieve the latent vectors <code>z</code> for any given <code>x</code> input.</p>
<h3 id="4-visualize-the-learned-cell-specific-representations-using-umap">4) Visualize the learned cell specific representations using UMAP<a class="headerlink" href="#4-visualize-the-learned-cell-specific-representations-using-umap" title="Permanent link">&para;</a></h3>
<p>You can use <code>scanpy</code> to visualize the learned cell specific representations using UMAP.
Does the latent space appear coherent? Can you validate whether the latent space preserves any prior annotations expected to dominate the signal?</p>
<h3 id="5-compare-against-decipher">5) Compare against Decipher<a class="headerlink" href="#5-compare-against-decipher" title="Permanent link">&para;</a></h3>
<p>We would now like to see how our simple autoencoder model compares to Decipher.</p>
<p>Install <code>decipher</code> by running:
<div class="language-bash highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>pip<span class="w"> </span>install<span class="w"> </span>scdecipher
</span></code></pre></div></p>
<p>Decipher has an API close to <code>scanpy</code>, with computational function in <code>decipher.tl</code> and plotting functions in <code>decipher.pl</code>.
You can fit a Decipher model to the data using the following code snippet:
<div class="language-python highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="kn">import</span> <span class="nn">decipher</span> <span class="k">as</span> <span class="nn">dc</span>
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="c1"># color by cell type (called &#39;str_labels&#39; in the pbmc dataset)</span>
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="n">model</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">decipher_train</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">plot_every_k_epochs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">plot_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;str_labels&#39;</span><span class="p">))</span>
</span></code></pre></div></p>
<p>You should be able to train the model and retrieve a similar latent representation.
Visualize this representation and compare it against the one from your autoencoder.</p>
<p>How do they differ?</p>
<p>Optional:</p>
<ul>
<li>Decode a series of points across a data-dense region of the latent representations of the auto-encoder and Decipher.</li>
<li>Then, for each gene (or a select few genes), plot the trend in gene expression values corresponding to the series of points.</li>
<li>Do they appear smooth or discontinuous?</li>
<li>Are there correlations between certain genes?</li>
<li>Are there sudden shifts in gene expression that correspond with annotation changes?</li>
</ul>
<h2 id="problem-3-implement-your-own-model">Problem 3 - Implement your own model<a class="headerlink" href="#problem-3-implement-your-own-model" title="Permanent link">&para;</a></h2>
<p>Now you can implement your own model on your own data. Good luck!</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Maintained by <a href="https://github.com/ANazaret">Achille</a>.
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/anazaret/iicd-workshop-2024" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://pypi.org/project/iicd-workshop-2024" target="_blank" rel="noopener" title="pypi.org" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6zM286.2 404c11.1 0 20.1 9.1 20.1 20.3 0 11.3-9 20.4-20.1 20.4-11 0-20.1-9.2-20.1-20.4.1-11.3 9.1-20.3 20.1-20.3zM167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4zm-6.7-142.6c-11.1 0-20.1-9.1-20.1-20.3.1-11.3 9-20.4 20.1-20.4 11 0 20.1 9.2 20.1 20.4s-9 20.3-20.1 20.3z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
        <script src="../javascripts/katex.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>